### PR TITLE
Allow port to be selected in spanning tree tool

### PR DESCRIPTION
**Pull request by jackgerrits**
In order to parallelize tests, they need to run on different ports. This allows us to supply the spanning tree tool with a custom port.

### CODE REVIEW

File: cluster/spanning_tree_main.cc [hunk 1]
```diff
@@ -17,33 +17,70 @@ int getpid() { return (int)::GetCurrentProcessId(); }
...
-#include <errno.h>
 
 using namespace VW;
 
+#include <boost/program_options.hpp>
+namespace po = boost::program_options;
+
+void usage(const po::options_description& desc)
+{
+  std::cout << "usage: spanning_tree [--port,-p number] [--nondaemon] [--help,-h] [pid_file]" << std::endl;
```
Comment (lalo) [inline cluster/spanning_tree_main.cc:28]: is the boost print help message easy to incorporate?

Comment (jackgerrits) [reply cluster/spanning_tree_main.cc:28]: Thats what line 29 is. It doesn't give you one in this form though and it is important to signal the final positional arg

File: cluster/spanning_tree_main.cc [hunk 2]
```diff
@@ -17,33 +17,70 @@ int getpid() { return (int)::GetCurrentProcessId(); }
...
+  std::cout << desc << std::endl;
+}
+
 int main(int argc, char* argv[])
 {
-  if (argc > 2)
+  int port = 26543;
```
Comment (olgavrou) [inline cluster/spanning_tree_main.cc:34]: has this always been the default port?

Comment (jackgerrits) [reply cluster/spanning_tree_main.cc:34]: yep, its in the spanning tree constructor. its also hard coded into vw

File: cluster/spanning_tree_main.cc [hunk 3]
```diff
@@ -17,33 +17,70 @@ int getpid() { return (int)::GetCurrentProcessId(); }
...
+  try
+  {
+    po::store(po::command_line_parser(argc, argv).options(all_options).positional(pos).run(), vm);
+    po::notify(vm);
+  }
+  catch (std::exception& e)
+  {
+    std::cout << argv[0] << ": " << e.what() << std::endl << std::endl;
+    usage(desc);
+    exit(1);
```
Comment (olgavrou) [inline cluster/spanning_tree_main.cc:62]: does this need to exit? can it `return 1`?

Comment (jackgerrits) [reply cluster/spanning_tree_main.cc]: Return makes sense

File: cluster/spanning_tree_main.cc (hunk 2 above)
Comment (olgavrou) [reply cluster/spanning_tree_main.cc:34]: should it live as a constexpr somewhere then and be re-used?

Comment (jackgerrits) [reply cluster/spanning_tree_main.cc:34]: This binary does not depend on VW at all. Maybe in future. Common lib constants or something?

Comment (olgavrou) [reply cluster/spanning_tree_main.cc:34]: mmm ok I see

**Review by olgavrou — APPROVED** [review]

**Pull request merged**